cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -167,9 +168,14 @@ static inline Result<csv::ReadOptions> GetReadOptions(
 static inline Future<std::shared_ptr<csv::StreamingReader>> OpenReaderAsync(
     const FileSource& source, const CsvFileFormat& format,
     const std::shared_ptr<ScanOptions>& scan_options, Executor* cpu_executor) {
+#ifdef ARROW_WITH_OPENTELEMETRY
+  auto tracer = arrow::internal::tracing::GetTracer();
+  auto span = tracer->StartSpan("arrow::dataset::CsvFileFormat::OpenReaderAsync");
```

Comment:
Maybe we can use the START_SPAN macro for this?

Reply:
I'd like to, but it results in errors such as 
```
arrow/cpp/src/arrow/dataset/file_csv.cc:205:7: error: use of deleted function ‘arrow::util::tracing::Span::Span(const arrow::util::tracing::Span&)’
  205 |       [=](const Status& err) -> Result<std::shared_ptr<csv::StreamingReader>> {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  206 | #ifdef ARROW_WITH_OPENTELEMETRY
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  207 |         arrow::internal::tracing::MarkSpan(err, span.get());
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  208 |         span->End();
      |         ~~~~~~~~~~~~
  209 | #endif
      | ~~~~~~ 
  210 |         return err.WithMessage("Could not open CSV input source '", path, "': ", err);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  211 |       });
      |       ~
```
Because I need to create a span to use these macros:
```
util::tracing::Span span;
  START_SPAN(span, "arrow::dataset::CsvFileFormat::OpenReaderAsync");
```
This also happens in other places

Reply:
seems `util::tracing::Span` is not copyable while OTel's span is (makes sense, since OTel's span is a shared_ptr)

Reply:
And we can't yet use move in capture expressions :(

Reply:
Just to point out here that we should investigate removing the `util::tracing::Span` wrapper class, and then we should be able to do this with a `START_SPAN` here and in all of the other places